### PR TITLE
[c10d] Use TCPStore to record NCCL timeout and dump debug info

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115404
* #115403
* __->__ #115226

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @wz337 @tianyu-l @wconstab @yf225 @kiukchung @lucasllc

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -1227,7 +1227,21 @@ std::string ProcessGroupNCCL::getNCCLWatchdogDebugInfo() {
   return retrieveDesyncReport(store_, "NCCL", rank_, size_);
 }
 
-void ProcessGroupNCCL::workCleanupLoop() {
+bool ProcessGroupNCCL::shouldAutoDumpWorkTimeOut(WorkNCCL& work) {
```
Comment (zdevito) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1230]: I'm not following what the intent of this function is.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -1247,6 +1261,13 @@ void ProcessGroupNCCL::workCleanupLoop() {
       auto& work = *it;
       work.checkAndSetException();
       bool timedOut = work.checkTimeout();
+      c10::optional<std::thread> dumpingDebugInfo;
```
Comment (zdevito) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1264]: A case we need to handle is when a process is the cause of a timeout: it hasn't started work, so others are blocked. In this case the workMetaList_ will be empty, and this code won't run.

I think we likely need a check where `heartbeat_++` is: 

```
if (last_time_we_removed_something_from_metaworkList - now > watchdog_timeout) {
   if (last_time_we_polled_for_dump_timeout > heartbeat_timeout) {
       if (timeout_dump is set in store) {
            <do a local dump> 
       } 
   }
}
```

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -1227,7 +1227,21 @@ std::string ProcessGroupNCCL::getNCCLWatchdogDebugInfo() {
...
-void ProcessGroupNCCL::workCleanupLoop() {
+bool ProcessGroupNCCL::shouldAutoDumpWorkTimeOut(WorkNCCL& work) {
+  if (!dumpOnTimeout_)
+    return false;
+  auto timeout = std::chrono::milliseconds(heartbeatTimeoutInSec_ * 1000);
+  return work.checkTimeout(c10::optional<std::chrono::milliseconds>(timeout));
+}
+
+void ProcessGroupNCCL::checkAndSetStore() {
+  if (!store_->check({std::string(TIMEOUT_DUMP)})) {
```
Comment (zdevito) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1238]: Is there a need to check TIMEOUT_DUMP here vs just setting it? It is 1/2 the messages to just set and would be a race anyway.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 4]
```diff
@@ -1247,6 +1261,13 @@ void ProcessGroupNCCL::workCleanupLoop() {
       auto& work = *it;
       work.checkAndSetException();
       bool timedOut = work.checkTimeout();
+      c10::optional<std::thread> dumpingDebugInfo;
+
+      if (shouldAutoDumpWorkTimeOut(work)) {
+        checkAndSetStore();
+        // Store debug info to storage. (By default to local disk)
+        dumpingDebugInfo = tryWriteDebugInfo();
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1269]: i think i am gonna land my PR first (well, i already told the bot but we'll see).  So you may want to rebase on mine to get used to the new future thing

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp [hunk 5]
```diff
@@ -659,10 +661,14 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   // Watchdog's inside loop.
   // Takes care of cleaning up completed work, and aborting upon failure or
   // timeout.
-  void workCleanupLoop();
+  void watchdogLoopHandler();
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:664]: nit: do the rename in a separate PR.  it'd be annoying to revert/reland the rename and its better to get ppl aligned and land it quickly

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (fduwjj) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1230]: Well, I guess there are two different ways:
1. Each rank keeps polling the TCPStore and check whether the flag has been sent. 
2. Check whether one event is old enough that we need to dump the debug info.

Previously I want to go with #1 but looks like #2 is better so that we don't have a lot of traffic to TCPStore?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 3 above)
Comment (fduwjj) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1238]: Right, we can just set, because we don't care about race condition here.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 4 above)
Comment (fduwjj) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1269]: Yes, I definitely want to do rebase once.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (fduwjj) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1264]: ok, you are talking about the case when the heartbeat is still increasing, but somehow the `workMetaList_` keeps being empty since last time removed?

I want to confirm one thing, do we perform the poll everytime we loop through each Work?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (zdevito) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1230]: Basically we want to do (1), but with the optimization that we do not poll the TCPStore when we know it is impossible for another rank to have timed out (`last_time_we_removed_something_from_metaworkList - now < watchdog_timeout`)

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (zdevito) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1264]: The if statement is _outside_ of the loop over work. The idea that we should poll once very heartbeat_timeout seconds only if we haven't seen any work start or complete in the last watchdog_timeout seconds.  This is essentially just an optimization of always polling the store every heartbeat_timeout seconds: we know no one could have timed out if we have recently finished some nccl work, so there is no need to check the store.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (wconstab) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1230]: i guess the only exception here could be if we created a sub-pg on some ranks and no equivalent sub-pgs covering other ranks, then other ranks could observe 'nothing is happening' for a while the trainer really is progressing?

Comment (zdevito) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp]: they could, but it would have to be a long time of nothing happening (like 20 minutes for the timeout), and all that would happen would be that every 2 seconds they would poll.  Once we are talking about a subset of process groups, who stay inactive for 20 minutes, I am not concerned that their polling would affect overall performance.

Comment (fduwjj) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp]: currently monitor time out is just 2 mins and watchdog timeout is 1 sec. Do you think we might need to make the timeout longer? Also, I updated the code based on my understanding of all comments and feedback. Do you mind taking a look again?

I will do the rebase later (it should be straightforward). And for test, I am still thinking how we can test this case? I mean this changes is just to further ensure we dump the flight recorder and it's kind of hard to mimic the situation when there is no add or erase.. Maybe I should make rank 1 sleep for a while?

**Comment by fduwjj** [comment]
Aside from rebase, which I will do it later, addressed most of feedback and wondering if this now looks in the right direction or not.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (wconstab) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp]: actually, i think the test i added for dumping on timeout is almost exactly what you need.  You can just customize it slightly so that instead of expecting rank1 to exit cleanly, you expect it also crashes and produces a separate flightrecorder file.  You'll also have to make sure rank1 stays alive long enough to get killed.

https://github.com/pytorch/pytorch/pull/115176/files#diff-9f639571a250cffbe9cded7d2fbb5ad6311e4be9c0c7610e5ba85930806e7f38

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 6]
```diff
@@ -1242,6 +1245,24 @@ void ProcessGroupNCCL::workCleanupLoop() {
...
+    // we haven't polled for `heartbeat_timeout` seconds and there haven't
+    // any work added or removed for `watchdog_timeout` seconds.
+    if (dumpOnTimeout_) {
+      auto currentTimepoint = std::chrono::steady_clock::now();
+      if (std::chrono::duration_cast<std::chrono::milliseconds>(
+              currentTimepoint - lastWorkListUpdateTime_) >=
+              std::chrono::milliseconds(kWatchdogThreadSleepMillis) &&
+          std::chrono::duration_cast<std::chrono::seconds>(
+              currentTimepoint - lastTimePollStore) >=
+              std::chrono::seconds(heartbeatTimeoutInSec_)) {
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1258]: i'm unsure why you need to keep track of both lastWorkListUpdateTime_ and lastTimePollStore, and why you use two separate thresholds (kWatchdogThreadSleepMillis and heartbeatTimeoutInSec_).

The 'in english' version of this is really just "has it been a while since we've observed local progress?  if so, lets ping the store and see if anyone else timed out".  Any reason we can't do that with only the `lastWorkListUpdateTime_` portion of the code?

Also, what's your reasoning for how long is long enough?

**Comment by wconstab** [comment]
other than my comment above i think this looks better

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 6 above)
Comment (fduwjj) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1258]: Well, I guess the logic here is to optimize the poll so that we don't frequently read from Store. So for polling we use `heartbeatTimeoutInSec_` so that we wait at least that amount of time to check for the next time. And also if we see work is being added or removed from the list, it's not time yet, either. Well, I think one thing to simplify the logic here is to use only `lastWorkListUpdateTime_` and `heartbeatTimeoutInSec_`. cc: @zdevito to see if he has any comments on my proposal.

I don't have an answer to your second question but it does not matter right? We at least wait for some time (which in this case 2 mins) so that we don't query Store crazily. Let's say in 32K's scenario, query every 2 min, you get around 300 query/sec, which is reasonably low?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (wconstab) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp]: You could probably make rank1 wait for a signal from the parent which it will never recv.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 6 above)
Comment (wconstab) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1258]: yea there should be 2 constraints. First we want to not overwhelm the store, Second we want to dump soon enough to avoid being killed before we dump.  But i couldn't say what the values are for either of those constraints.  Maybe 2 min is fine.

**Comment by fduwjj** [comment]
OK, rebased and test added. Ping on review again.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 6 above)
Comment (fduwjj) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1330]: So what is the conclusion here? I can change the logic and I think this is the only blocker which prevent this PR from merging I guess?

Comment (wconstab) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1330]: my bad. i agree we need both conditions.  (would be nice if we could clean up the way it looks a bit )

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 7]
```diff
@@ -1314,6 +1317,24 @@ void ProcessGroupNCCL::workCleanupLoop() {
...
+        lastTimePollStore = currentTimepoint;
+        if (store_->check({std::string(TIMEOUT_DUMP)}) && !optAsyncDebugDump) {
+          optAsyncDebugDump = launchAsyncDebugDump();
+        }
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1334]: should we kill the process here? we know someone else timed out

**Review by wconstab — APPROVED** [review]
if you fix the nits i am good with this

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 7 above)
Comment (fduwjj) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1334]: yes, will add a wait_for and throw an exception here.

**Comment by fduwjj** [comment]
All nits fixed..

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 8]
```diff
@@ -1354,6 +1357,31 @@ void ProcessGroupNCCL::workCleanupLoop() {
...
+          timeSinceLastPollStore >= heartbeatTimeoutInSec_ * 1000) {
+        lastTimePollStore = currentTime;
+        if (store_->check({std::string(TIMEOUT_DUMP)}) && !optAsyncDebugDump) {
+          optAsyncDebugDump = launchAsyncDebugDump();
+          optAsyncDebugDump->wait_for(
+              std::chrono::milliseconds(kWatchdogThreadSleepMillis * 30));
+          const auto exitMsg = c10::str(
+              "[Rank ",
+              rank_,
+              "] NCCL watchdog thread detected timeout from Store");
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1378]: this message could be clearer.  (think of a user, they don't know what Store is or why it would be timing out.

More helpful would be to explain that some other rank's watchdog detected a timeout and notified us so we're dumping and aborting.

**Comment by wconstab** [comment]
I put a few more nits, but don't stop the merge.. maybe just fix in another PR

File: test/distributed/test_c10d_nccl.py [hunk 9]
```diff
@@ -3846,6 +3847,55 @@ def test_timeout_dumps(self):
             torch.cuda.synchronize()
 
 
+class NCCLTraceTestTimeoutDumpOnIdleRanks(NCCLTraceTestDumpOnTimeoutBase):
+    def _check_return_codes(self, elapsed_time):
+        # the base test infra assumes processes exit with matching return codes,
+        # but we want rank0 to abort and rank1 to exit cleanly in this test
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/test_c10d_nccl.py:3853]: Yeah.. If it was happy with (-6, 0), I wouldn't bother doing it. Also why I had to create `NCCLTraceTestDumpOnTimeoutBase` is because if I directly inherit `NCCLTraceTestDumpOnTimeout`, when using parametrize python don't allow that. 

Let me change the comment in another PR.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 8 above)
Comment (fduwjj) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1378]: sure.. That is a good point, let me change it.
